mm2src/coins/coin_errors.rs

```diff
@@ -0,0 +1,216 @@
+    InternalError(String),
+    InvalidPaymentTxData(String),
+    InvalidInput(String),
+    ScriptHashTypeNotSupported(String),
+    SPVError(String),
+    TransportError(String),
+    UnexpectedDerivationMethod(UnexpectedDerivationMethod),
+    UnexpectedErc20PaymentData(String),
+    UnexpectedPaymentState(String),
+    UnexpectedPaymentTx(String),
+    ValidateHtlcError(String),
+    #[display(fmt = "Payment tx token_addr arg {:?} is invalid, expected {:?}", found, expected)]
+    WrongTokenAddress {
+        found: String,
+        expected: String,
+    },
+    #[display(fmt = "Payment tx value arg {:?} is invalid, expected {:?}", found, expected)]
+    WrongValue {
+        found: String,
+        expected: String,
+    },
+    #[display(fmt = "Payment tx time_lock arg {:?} is invalid, expected {:?}", found, expected)]
+    WrongTimeLock {
+        found: String,
+        expected: String,
+    },
+    #[display(fmt = "Invalid 'swap_id' {:?}, expected {:?}", found, expected)]
+    WrongSwapId {
+        found: String,
+        expected: String,
+    },
+    #[display(fmt = "Payment tx receiver arg {:?} is invalid, expected {:?}", found, expected)]
+    WrongReceiver {
+        found: String,
+        expected: String,
+    },
+    #[display(fmt = "Payment tx secret_hash arg {:?} is invalid, expected {:?}", found, expected)]
+    WrongSecretHash {
+        found: String,
+        expected: String,
+    },
+    #[display(fmt = "Payment tx {:?} was sent to wrong address, expected {:?}", found, expected)]
+    WrongReceiverAddress {
+        found: String,
+        expected: String,
+    },
+    #[display(fmt = "Payment tx {:?} was sent from wrong address, expected {:?}", found, expected)]
+    WrongSenderAddress {
+        found: String,
+        expected: String,
+    },
+}
+
+impl From<ethabi::Error> for ValidatePaymentError {
+    fn from(err: ethabi::Error) -> Self { Self::InvalidPaymentTxData(err.to_string()) }
+}
+
+impl From<rlp::DecoderError> for ValidatePaymentError {
+    fn from(err: rlp::DecoderError) -> Self { Self::InvalidPaymentTxData(err.to_string()) }
+}
+
+impl From<keys::Error> for ValidatePaymentError {
+    fn from(err: keys::Error) -> Self { Self::InternalError(err.to_string()) }
+}
+
+impl From<web3::Error> for ValidatePaymentError {
+    fn from(err: web3::Error) -> Self { Self::TransportError(err.to_string()) }
+}
+
+impl From<NumConversError> for ValidatePaymentError {
+    fn from(err: NumConversError) -> Self { Self::InternalError(err.to_string()) }
+}
+
+impl From<SPVError> for ValidatePaymentError {
+    fn from(err: SPVError) -> Self { Self::SPVError(err.to_string()) }
+}
+
+impl From<serialization::Error> for ValidatePaymentError {
+    fn from(err: serialization::Error) -> Self { Self::InternalError(err.to_string()) }
+}
+
+impl From<ScriptHashTypeNotSupported> for ValidatePaymentError {
+    fn from(err: ScriptHashTypeNotSupported) -> Self { Self::ScriptHashTypeNotSupported(err.to_string()) }
+}
```

Comment:
maybe not. I can actually implement `From<ScriptHashTypeNotSupported>` in `QTUM` mod alone and It won't have any of the side effects you listed above. but since it's being used here
```rust
let expected_receiver = qtum::contract_addr_from_utxo_addr(my_address).mm_err(|e| ValidatePaymentError::Internal(e.to_string()))?;
``` 

then I can do it manually.

Reply:
Implementing traits in private submodules occurs at the crate level.
For example,
```rust
struct MyFoo;

struct MyBar;

mod foo {
    use crate::{MyBar, MyFoo};

    impl From<MyFoo> for MyBar {
        fn from(_: MyFoo) -> Self {
            println!("from MyFoo for MyBar");
            MyBar
        }
    }
}

fn main() {
    //
    let x = MyFoo;
    let y = MyBar::from(x);
}
```
This will work.